cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java

```diff
@@ -534,47 +546,126 @@ public Void call() throws Exception {
+      Class<? extends Mapper> mapper = mapperInput.getMapper();
+      if (mapper != null) {
+        mapperClasses.add(mapper);
+      }
     }
 
-    if (provider != null) {
-      Configuration jobConf = job.getConfiguration();
-      ConfigurationUtil.setAll(provider.getInputFormatConfiguration(), jobConf);
-      jobConf.set(Job.INPUT_FORMAT_CLASS_ATTR, provider.getInputFormatClassName());
+    if (mapperClasses.size() < numInputs) {
+      // means one of the inputs is going to default to use the mapper assigned on the job,
+      // so also check the job's mapper class
+      if (!isProgrammaticConfig(job.getConfiguration(), Job.MAP_CLASS_ATTR)) {
+        // mapper class was not explicitly set on the job (reduce only job)
+        return null;
+      }
+      mapperClasses.add(job.getMapperClass());
+    }
+
+    // this shouldn't ever happen, because we know there's at least 1 input configured
+    Preconditions.checkArgument(!mapperClasses.isEmpty());
+
+    Class<? extends Mapper> firstMapperClass = mapperClasses.get(0);
+    TypeToken<Mapper> firstType = resolveClass(firstMapperClass, Mapper.class);
+    Type[] firstTypeArgs = ((ParameterizedType) firstType.getType()).getActualTypeArguments();
+    Class<?> firstMapOutputKey = TypeToken.of(firstTypeArgs[2]).getRawType();
+    Class<?> firstMapOutputValue = TypeToken.of(firstTypeArgs[3]).getRawType();
+
+    // compare the remainder of the configured mappers (if any) with the first
+    for (int i = 1; i < mapperClasses.size(); i++) {
+      Class<? extends Mapper> mapperClass = mapperClasses.get(i);
+      TypeToken<Mapper> type = resolveClass(mapperClass, Mapper.class);
+      Type[] typeArgs = ((ParameterizedType) type.getType()).getActualTypeArguments();
+      Class<?> mapOutputKey = TypeToken.of(typeArgs[2]).getRawType();
+      Class<?> mapOutputValue = TypeToken.of(typeArgs[3]).getRawType();
+      if (!firstMapOutputKey.equals(mapOutputKey) || !firstMapOutputValue.equals(mapOutputValue)) {
+        throw new IllegalArgumentException(
+          String.format("Type mismatch in output type of mappers: %s and %s. " +
+                          "Map output key types: %s and %s. " +
+                          "Map output value types: %s and %s.",
+                        firstMapperClass, mapperClass,
+                        firstMapOutputKey, mapOutputKey,
+                        firstMapOutputValue, mapOutputValue));
+      }
+    }
+    return firstType;
+  }
+
+  /**
+   * Sets the configurations used for inputs.
+   */
+  private void setInputsIfNeeded(Job job) throws IOException {
+    Map<String, MapperInput> mapperInputs = context.getMapperInputs();
+
+    if (mapperInputs.size() == 0) {
+      // user is not going through our APIs to add input; leave the job's input format to user
+      return;
+    }
+
+    // keep track of which mapper was defined directly on the job, to use as the default
+    MultipleInputs.setRootMapper(job);
```

Comment:
I think it's better call it `defaultMapper`. Root sounds more like the one at the top who does the delegation.